src/input.rs

```diff
@@ -46,11 +61,50 @@ impl<I: Iterator<Item = Result<char, io::CharsError>>> Iterator for Keys<I> {
     fn next(&mut self) -> Option<Result<Key, io::CharsError>> {
         Some(match self.chars.next() {
             Some(Ok('\x1B')) => Ok(match self.chars.next() {
+                Some(Ok('O')) => match self.chars.next() {
+                    Some(Ok('P')) => Key::F(1),
+                    Some(Ok('Q')) => Key::F(2),
+                    Some(Ok('R')) => Key::F(3),
+                    Some(Ok('S')) => Key::F(4),
+                    _ => Key::Invalid,
+                },
                 Some(Ok('[')) => match self.chars.next() {
                     Some(Ok('D')) => Key::Left,
                     Some(Ok('C')) => Key::Right,
                     Some(Ok('A')) => Key::Up,
                     Some(Ok('B')) => Key::Down,
+                    Some(Ok('H')) => Key::Home,
+                    Some(Ok('F')) => Key::End,
+                    Some(Ok(c @ '1' ... '6')) => match self.chars.next() {
+                        Some(Ok('~')) => match c {
+                            '1' | '7' => Key::Home,
+                            '2' | '8' => Key::Insert,
+                            '3' => Key::Delete,
+                            '4' => Key::End,
+                            '5' => Key::PageUp,
+                            '6' => Key::PageDown,
+                            _ => Key::Invalid,
+                        },
+                        Some(Ok(k @ '0' ... '9')) => match self.chars.next() {
+                            Some(Ok('~')) => match (c, k) {
+                                ('1', '1') => Key::F(1),
```

Comment:
I'm down with using a formula instead of that, but this one doesn't work, notice how we go from `('1', '5')` to `('1', '7')` and from `('2', '1')` to `('2', '3')`